File: binary-tree/js/BST.js
class Node {
  constructor(value, left = null, right = null) {
    this.value = value;
    this.left = left;
    this.right = right;
  }
}

class BTS {
  constructor(value) {
    this.root = new Node(value);
  }

  insert(value) {
    let currentNode = this.root;

    while (true) {
      if (currentNode.value < value) {
        if (!currentNode.right) {
          let node = new Node(value);
          currentNode.right = node;
          break;
        }

        currentNode = currentNode.right;
      } else {
        if (!currentNode.left) {
          let node = new Node(value);
          currentNode.left = node;
          break;
        }

        currentNode = currentNode.left;
      }
    }
  }

  search(target) {
    let node = this.root;

    while (node) {
      if (node.value === target) {
        return `${target} is found`;
      }

      if (node.value < target) {
        if (!node.right) {
          return "Not Found";
        }
        node = node.right;
      }

      if (node.value > target) {
        if (!node.left) {
          return "Not Found";
        }
        node = node.left;
      }
    }
  }

  BFS() {
    const arr = [this.root];

    while (arr.length != 0) {
      let currentNode = arr.shift();
      console.log(currentNode.value);

      if (currentNode.left) {
        arr.push(currentNode.left);
      }

      if (currentNode.right) {
        arr.push(currentNode.right);
      }
    }
  }

  DFSInOrderTree(node) {
    if (!node) {
      return;
    }
    this.DFSInOrderTree(node.left);
    console.log(node.value);
    this.DFSInOrderTree(node.right);
  }

  DFSPreOrderTree(node) {
    if (!node) {
      return;
    }
    console.log(node.value);
    this.DFSPreOrderTree(node.left);
    this.DFSPreOrderTree(node.right);
  }

  DFSPostOrderTree(node) {
    if (!node) {
      return;
    }

    this.DFSPostOrderTree(node.left);
    this.DFSPostOrderTree(node.right);
    console.log(node.value);
  }

  isValidBST(node = null, minimum = -Infinity, maximum = Infinity) {
    if (!node) {
      return true;
    }

    if (node.value < minimum || node.value > maximum) {
      return false;
    }

    return (
      this.isValidBST(node.left, minimum, node.value) &&
      this.isValidBST(node.right, node.value, maximum)
    );
  }

  predessor(root) {
    let node = root.left;

    while (true) {
      if (!node.right) {
        console.log(node.value);
        break;
      }

      node = node.right;
    }
  }

  successor(root) {
    let node = root.right;

    while (true) {
      if (!node.left) {
        console.log(node.value);
        break;
      }
      node = node.left;
    }
  }
}

const bts = new BTS(10);
bts.insert(9);
bts.insert(11);
bts.insert(8);
bts.insert(12);
bts.search(1);
console.log(bts.isValidBST(bts.root));
bts.predessor(bts.root);
bts.successor(bts.root);
console.log(bts.search(1));
bts.BFS();
bts.DFSInOrderTree(bts.root);
bts.DFSPreOrderTree(bts.root);
bts.DFSPostOrderTree(bts.root);

console.log(bts);

File: binary-tree/py/BST.py

class Node:
    def __init__(self, val = None, left = None , right = None):
      self.val = val
      self.left = left
      self.right = right

    


class BST:
   def __init__(self,val = None):
     self.root = Node(val) if val is not None else None
   


   def insert(self , val):
      node = self.root

      if node is None:
        self.root = Node(val)
        return 

     
      while True:
        if node.val > val:
          if node.left is None:
            node.left = Node(val)
            break
        
          node = node.left 
        else:
          if node.right is None:
            node.right = Node(val)
            break
        
          node = node.right

   def search(self,val):
       node = self.root

       if node is None:
         print("there is no val in BST")
         return
       
       while node:
         if node.val == val:
           return f"{val} is found"
         
         if node.val > val:
           if node.left is None:
             return f"{val} not found"
           node = node.left

         if node.val < val:
           if node.right is None:
             return f"{val} not found"
           node = node.right

   def BFS(self):
     
     node = self.root
     arr = [node]

     if node is None:
       return "there is no val in BTS"
     
     while len(arr) != 0:
        first_node = arr.pop(0)
        print(first_node.val)

        if first_node.left is not None:
          arr.append(first_node.left)
        
        if first_node.right is not None:
          arr.append(first_node.right)

   def DFS_in_order_tree(self,node):
     if node is None:
       return
     
     self.DFS_in_order_tree(node.left)
     print(node.val)
     self.DFS_in_order_tree(node.right)
  
   def DFS_pre_order_tree(self,node):
     
     if node is None:
       return
     
     print(node.val)
     self.DFS_pre_order_tree(node.left)
     self.DFS_pre_order_tree(node.right)

   def DFS_post_order_tree(self,node):
     
     if node is None:
       return
     
     self.DFS_post_order_tree(node.left)
     self.DFS_post_order_tree(node.right)
     print(node.val)

   def predessor(self,root):
    node = root.left

    while True:
      if node.right is None:
        print(node.val)
        break

      node = node.right

   def successor(self,root):
    node = root.right

    while True:
      if node.left is None:
        print(node.val)
        break

      node = node.left

   def is_valid_bst(self,node = None , minimum=float('-inf'), maximum=float('inf')):
    if node is None:
      return True
    
    if node.val < minimum or node.val > maximum:
      return False
    
    return (
        self.is_valid_bst(node.left, minimum, node.val) and
        self.is_valid_bst(node.right, node.val, maximum)
    )
     
         


bst = BST()
bst.insert(10)
bst.insert(9)
bst.insert(11)
bst.insert(20)
bst.insert(12)
bst.predessor(bst.root)
bst.successor(bst.root)
print(bst.is_valid_bst(bst.root))
# bst.BFS()
# print(bst.search(42))
# bst.DFS_in_order_tree(bst.root)
# bst.DFS_pre_order_tree(bst.root)
# bst.DFS_post_order_tree(bst.root)


